src/pkg/task/sweep_manager.go

```diff
@@ -0,0 +1,189 @@
+	ormer, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	vendorIDs, err := sm.listVendorIDs(ctx, vendorType)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list vendor ids for vendor type %s", vendorType)
+	}
+
+	// execIDs stores the result
+	var execIDs []int64
+	for _, vendorID := range vendorIDs {
+		maxStartTime, err := sm.getCandidateMaxStartTime(ctx, vendorType, vendorID, retainCnt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get candidate max start time, vendor type: %s, vendor id: %d", vendorType, vendorID)
+		}
+		// continue if no max start time got that means no candidate executions
+		if maxStartTime == nil {
+			continue
+		}
+		// candidate criteria
+		// 1. exact vendor type & vendor id
+		// 2. start_time is before the max start time
+		// 3. status is the final state
+		// count the records for pagination
+		sql := `SELECT COUNT(1) FROM execution WHERE vendor_type = ? AND vendor_id = ? AND start_time < ? AND status IN (?,?,?)`
+		totalOfCandidate := 0
+		params := []interface{}{
+			vendorType,
+			vendorID,
+			maxStartTime.Format(timeFormat),
+			// final status should in Error/Success/Stopped
+			job.ErrorStatus.String(),
+			job.SuccessStatus.String(),
+			job.StoppedStatus.String(),
+		}
+		if err = ormer.Raw(sql, params...).QueryRow(&totalOfCandidate); err != nil {
+			return nil, errors.Wrapf(err, "failed to count candidates, vendor type: %s, vendor id: %d", vendorType, vendorID)
+		}
+		// n is the page count of all candidates
+		n := totalOfCandidate / defaultPageSize
+		if totalOfCandidate%defaultPageSize > 0 {
+			n = n + 1
+		}
+
+		sql = `SELECT id FROM execution WHERE vendor_type = ? AND vendor_id = ? AND start_time < ? AND status IN (?,?,?)`
+		// default page size is 100000
+		q2 := &q.Query{PageSize: int64(defaultPageSize)}
+		for i := n; i >= 1; i-- {
+			q2.PageNumber = int64(i)
+			// should copy params as pagination will append the slice
+			paginationParams := make([]interface{}, len(params))
+			copy(paginationParams, params)
+			paginationSQL, paginationParams := orm.PaginationOnRawSQL(q2, sql, paginationParams)
+			ids := make([]int64, 0, defaultPageSize)
+			if _, err = ormer.Raw(paginationSQL, paginationParams...).QueryRows(&ids); err != nil {
+				return nil, errors.Wrapf(err, "failed to list candidate execution ids, vendor type: %s, vendor id: %d", vendorType, vendorID)
+			}
+			execIDs = append(execIDs, ids...)
+		}
+	}
+
+	return execIDs, nil
+}
+
+func (sm *sweepManager) Clean(ctx context.Context, execIDs []int64) error {
+	ormer, err := orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+	// construct sql params
+	params := make([]interface{}, 0, len(execIDs))
+	for _, eid := range execIDs {
+		params = append(params, eid)
+	}
+	// delete tasks
+	sql := fmt.Sprintf("DELETE FROM task WHERE status_code = %d AND execution_id IN (%s)", finalStatusCode, orm.ParamPlaceholderForIn(len(params)))
```

Comment:
Is it possible replace it with this SQL?
```
DELETE FROM executions e1
WHERE NOT EXISTS (
  SELECT 1 FROM (
    SELECT id 
    FROM executions e2 
    WHERE e2.vendor_type = e1.vendor_type AND e2.status != 'Running'
    ORDER BY end_time DESC 
    LIMIT 50
  ) AS latest_executions
) WHERE e1.status != 'Running' AND e1.vendor_type = latest_executions.vendor_type
```

Reply:
We need to delete tasks before executions, otherwise the execution cannot be deleted as key referenced.